### PR TITLE
TAJO-1338: Defines RESTful API for Clients

### PR DESCRIPTION
**Pull request by ykrips**
Detail API designs can be found in the following link.
https://cwiki.apache.org/confluence/display/TAJO/TAJO+REST+API

### CODE REVIEW

**Comment by ykrips** [comment]
This patch aligned to current version of source code.

**Comment by ykrips** [comment]
Testcases in travis-ci failed, but testcases in apache build servers has been passed. Please see details in the following link.
https://builds.apache.org/job/PreCommit-TAJO-Build/695//testReport/

File: tajo-core/src/main/java/org/apache/tajo/ws/rs/resources/FunctionsResource.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  @Context
+  Application application;
+  
+  @PathParam("databaseName")
+  String databaseName;
+  
+  JerseyResourceDelegateContext context;
+  
+  private static final String databaseNameKeyName = "databaseName";
+  private static final String functionNameKeyName = "functionName";
```
Comment (blrunner) [inline tajo-core/src/main/java/org/apache/tajo/ws/rs/resources/FunctionsResource.java]: Currently, it looks like unused variable.

File: tajo-core/src/main/java/org/apache/tajo/ws/rs/requests/SubmitQueryRequest.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  @Expose private String query;
+  @Expose private Map<String, String> sessionVariables;
+  
+  public String getQuery() {
+    return query;
+  }
+  public void setQuery(String query) {
+    this.query = query;
+  }
+  public Map<String, String> getSessionVariables() {
```
Comment (blrunner) [inline tajo-core/src/main/java/org/apache/tajo/ws/rs/requests/SubmitQueryRequest.java]: SubmitQueryRequest::getSessionVariables and SubmitQueryRequest:: setSessionVariables seem to be unused methods.

**Comment by blrunner** [comment]
+1

Great! It looks good overall to me and I found that rest server response successfully on my laptop as follows.
```
$ curl -XGET 'http://localhost:26880/rest/databases'
{"databases":["default","information_schema"]}

$ curl -XGET 'http://localhost:26880/rest/databases/default/tables'
[{"tableName":"customer","external":true},{"tableName":"lineitem","external":true},{"tableName":"nation","external":true},{"tableName":"orders","external":true},{"tableName":"part","external":true},{"tableName":"partsupp","external":true},{"tableName":"supplier","external":true}]

$  curl -XGET 'http://localhost:26880/rest/cluster'
{"workers":[{"connectionInfo":{"id":-2105400716,"host":"192.168.219.106","peerRpcPort":28091,"pullServerPort":64176,"queryMasterPort":28093,"clientPort":28092,"httpInfoPort":28080},"diskSlots":0.0,"cpuCoreSlots":1,"memoryMB":1024,"usedDiskSlots":0.0,"usedMemoryMB":0,"usedCpuCoreSlots":0,"maxHeap":932184064,"freeHeap":365067488,"totalHeap":458227712,"numRunningTasks":0,"numQueryMasterTasks":0,"lastHeartbeatTime":1428221077181}]}
```
I leaved some trivial comments and I found unused methods in WorkerConnectionInfoResponse and WorkerResponse, too. Before you ship this, please check my comments. :)

**Comment by ykrips** [comment]
@blrunner ,
I have removed unused methods in my patch. Thanks for review.

**Comment by blrunner** [comment]
Thanks @ykrips 

I found the updated patch. Finally, could you trigger this for CI build or upload the patch to JIRA?

**Comment by ykrips** [comment]
No problems. I don't have any privileges on travis ci, so I will upload the patch to JIRA.

**Comment by blrunner** [comment]
I found the result of jenkins build as following:
https://builds.apache.org/job/PreCommit-TAJO-Build/709//testReport/

Ship it. :)

**Comment by ykrips** [comment]
@blrunner ,
Thanks you for quick review. This patch will commit into main branch.

**Pull request closed without merging**